test/unit/test-http-datasource.cpp

```diff
@@ -100,38 +101,63 @@ TEST_CASE("HttpDataSource", "[HttpDataSource]")
         REQUIRE(receivedTileCount == 1);
     }
 
-    SECTION("Stop the data source")
+    SECTION("Query mapget HTTP service")
     {
-        ds.stop();
-        REQUIRE(ds.isRunning() == false);
+        mapget::HttpService service;
+        service.add(std::make_shared<mapget::RemoteDataSource>("localhost", ds.port()));
+
+        service.go();
+
+        SECTION("Query through mapget HTTP service")
+        {
+            mapget::HttpClient client("localhost", service.port());
+
+            auto receivedTileCount = 0;
+            client.request(std::make_shared<mapget::LayerTilesRequest>(
+                               "Tropico",
+                               "WayLayer",
+                               std::vector<mapget::TileId>{{1234, 5678, 9112, 1234}},
+                               [&](auto&& tile) { receivedTileCount++; }))
+                ->wait();
+
+            REQUIRE(receivedTileCount == 4);
+            // One tile requested twice, so the cache was used.
+            REQUIRE(dataSourceRequestCount == 3);
+        }
+
+        SECTION("Trigger 400 responses")
+        {
+            mapget::HttpClient client("localhost", service.port());
+
+            auto unknownMapRequest = std::make_shared<mapget::LayerTilesRequest>(
+                "UnknownMap",
+                "WayLayer",
+                std::vector<mapget::TileId>{{1234}},
+                [&](auto&& tile) { /* TODO check something? */ });
```

Comment:
I think this TODO can be removed. But if you want to do something else here, you *could* ensure that the callback is never called by incrementing some counter and then checking that the counter is zero.